### DISCUSSION TITLE
How to inject GLSPActionDispatcher?

### DISCUSSION THREAD

**Discussion by Ginxss** [eclipse-glsp/glsp#888] [Q&A] [answered]
Hello,

I have a project which is similar to [this example project](https://github.com/eclipse-glsp/glsp-examples/tree/property-view-example/project-templates/java-emf-theia) with the property view. I want to tell the server to update the model when I detected changes in the property view. The problem is, I can't seem to find a way to get access to an ActionDispatcher in files that were bound in the frontend module, like the [widget provider and the property data service](https://github.com/eclipse-glsp/glsp-examples/tree/property-view-example/project-templates/java-emf-theia/glsp-client/tasklist-theia/src/browser/property-view).

To reproduce my problem, take the linked example project and add
```typescript
@inject(TYPES.IActionDispatcher)
protected actionDispatcher: GLSPActionDispatcher;
```
to custom-widget-provider.ts or custom-data-service.ts, which just throws the error
`ERROR _validateActiveBindingCount@http://localhost:3000/bundle.js`.

Injecting as a parameter in constructor() or a postConstruct() method does not work as well.

How do I inject values in those files and why are they different from the selection-data-service.ts for example, where the injection works?

Thanks

**Comment by ndoschek** [comment, marked as answer]
Hi @Ginxss,

The background of this problem is, that the property view is usually bound in the general frontend module and not diagram specific, as the action dispatcher is. An action dispatcher is available per diagram container, and can of course be unbound if no diagram was opened yet.
To overcome this, you could retrieve the actionDispatcher from the currently opened/active GLSP diagram widget in the custom data service for example:

```javascript
...
const widget = this.shell.activeWidget || this.shell.currentWidget;
if (widget instanceof GLSPDiagramWidget) {
   return widget.actionDispatcher;
}
...
```

I added another commit with a small working example to the previously used example branch, please see https://github.com/eclipse-glsp/glsp-examples/commit/70fdb1ab998c8e1e3051bf1b3266d1a21e44b2c3.
Remark: This is a very minimalistic example, which only fetches the name of the selected element without custom UI and type schemas or any editing support yet.

HTH and best wishes,
Nina


PS: In other projects we usually use the Model Server, an external source model management component, to fetch element property data.
This allows also to easily subscribe to changes and update the property view data.

**Reply by Ginxss** [reply]
@ndoschek Thanks!

If I may ask a follow up question: Since you brought up the Model Server, I'm still not entirely sure what its purpose is.
When making a cloud-based diagram editor, what benefits does it bring to have a Model Server next to the GLSP Server, aside from simplifying the communication of properties?
My current understanding is that I am going to be fine just remotely connecting to the Theia application which communicates with the GLSP Server directly.

**Reply by ndoschek** [reply]
When your use case targets a diagram editor only, then your current setup is surely sufficient.

In a use case where you would like to support different editing capabilities (such as a text-based and/or tree-based editor) the Model Server can manage this as an external source model management component.
If you are interested in how such a project looks like, I can point you to the [coffee-editor](https://www.eclipse.org/emfcloud/#coffeeeditoroverview), which supports multiple editing capabilities and uses also the Model Server.
